woocommerce/woocommerce-android
# [Tooling] Fix and automate upload of Play Store metadata during release builds

### Description

We recently discovered that the app description and related metadata were not up-to-date in the Play Store Console, and were not matching the latest copies we had in GlotPress and in `fastlane/metadata/android/**/*.txt`

The issue were twofold:

 - There was **no** automation in place to upload the `full_description.txt` etc. files to the Play Store Console as part of the release process.
    - Apparently, up until now, this was assumed that one would run the proper `fastlane supply` command manually when new metadata were expected to be uploaded, which can easily be missed (especially if the descriptions were updated by a developer without the Release Manager knowing that new Play Store descriptions had been updated in the repo and translated)
 - Also, we were missing the `fastlane/metadata/android/en-US` folder
    - which means metadata for English would not have been included in the automatic upload of the metadata by the `fastlane supply` (aka `fastlane upload_to_play_store`) action anyway.

### Solution 

- Added missing `en-US` folder and its `changelogs/` subfolder in `fastlane/metadata/android`, with a `.gitkeep` file to make sure it's kept around even if empty
- Fix `download_metadata_strings` for the `en-US` case, to make sure `.txt` files used as the source of truth for translation are copied into `fastlane/metadata/android/en-US` properly
- Finally, make our `build_and_upload_release` lane also upload metadata and release notes to PlayStore automatically, by turning `skip_upload_metadata` and `skip_upload_changelogs` to `false` during the upload of a new build to the production track.

### Testing instructions

 - [x] Run `bundle exec fastlane download_metadata_strings` and confirm that at the end of it, all the `*.txt` files in `fastlane/metadata/android/**.*.txt` match our expectation, and especially that it is now including the English source copy in `en-US/` subfolder
 - [ ] During the next release of WCAndroid on Friday, check that the Release Build CI step—which ends up calling `build_and_upload_google_play` which in turn is the one that calls `build_and_upload_release`—will properly update the metadata and release notes for you

> **Note**: I've only tested the first step, leaving the final test to you @ParaskP7  on Friday.

Actually you could consider testing that the update of the metadata work by using `upload_to_play_store` even before Friday (to avoid risk of last-minute surprises about translations too long or whatnot) by calling `upload_to_play_store` with just `skip_upload_metadata: false` and all the other `skip_*` to true, allowing you to just run that action to upload metadata even without having to wait for Friday or having to upload a new AAB or creating any new release.

You should be able to do that by adding a temporary dummy lane to your `Fastfile` like this:

```ruby
lane :upload_metadata_to_play_store do
  upload_to_play_store(
    package_name: APP_PACKAGE_NAME,
    skip_upload_aab: true,
    skip_upload_apk: true,
    skip_upload_metadata: false,
    skip_upload_changelogs: true,
    skip_upload_images: true,
    skip_upload_screenshots: true,
    json_key: UPLOAD_TO_PLAY_STORE_JSON_KEY,
    # Retain the build for v10.9.2 which is the latest one supporting Android 5/6
    version_codes_to_retain: [356],
  )
end
```

Then `bundle exec fastlane upload_metadata_to_play_store` and go check on Google Play Console to validate it worked. Or if the command failed on some locales saying the translation copies exceeded the max character limit, that will give you enough time to reach out to +i18n to ask for translators to provide shorter translations before Friday.

(@ParaskP7 I haven't done that test step on my end either, I'll let you run it and handle the follow-up with +18n if one is needed)

### Next Step

With this change, the release notes XML should already be automatically pre-populated by the `upload_to_play_store(…, skip_upload_changelog: false, …)` during CI release build.
Which means that we should later be able to get rid of:
 - The calls to `android_create_xml_release_notes` + `add_us_release_notes`, which are used to generate the `release_notes.xml` file
 - The `release_notes.xml` file itself, which we could delete as it won't be necessary nor generated anymore
 - The step in the WCAndroid release scenario that tells you to manuallly copy/paste the `release_notes.xml` into Play Store Console

@ParaskP7 I've left those two changes out of this PR just to make sure you have the opportunity to first do the next release of WCAndroid to confirm that the changes in this PR work. If you confirm they do work as expected, then feel free to create a subsequent PR next week to get rid of those (similar to what was done in https://github.com/wordpress-mobile/WordPress-Android/pull/16281)
